src/storage-controller/src/healthcheck.rs

```diff
@@ -0,0 +1,467 @@
+
+                    (row, 1)
+                },
+            )
+            .collect()
+    }
+
+    async fn append_updates(&mut self, updates: Vec<RawStatusUpdate>, type_: IntrospectionType) {
+        let source_status_history_id = *self
+            .introspection_ids
+            .lock()
+            .expect("poisoned lock")
+            .get(&type_)
+            .expect("source status history collection to be registered");
+
+        let new: Vec<_> = updates
+            .clone()
+            .into_iter()
+            .filter(
+                |r| match (&r.status_name, self.previous_statuses.get(&r.id).as_deref()) {
+                    // TODO(guswynn): Ideally only `failed` sources should not be marked as paused.
+                    // Additionally, dropping a replica and then restarting environmentd will
+                    // fail this check. This will all be resolved in:
+                    // https://github.com/MaterializeInc/materialize/pull/23013
+                    (new, Some(prev)) if new == "paused" && prev == "stalled" => false,
+                    (new, Some(prev)) => new != prev,
+                    _ => true,
+                },
+            )
+            .collect();
+
+        self.collection_manager
+            .append_to_collection(
+                source_status_history_id,
+                Self::pack_status_updates(new.clone()),
+            )
+            .await;
+
+        self.previous_statuses
+            .extend(new.into_iter().map(|r| (r.id, r.status_name)));
+    }
+
+    /// Appends updates for sources to the appropriate managed status collection
+    ///
+    /// # Panics
+    ///
+    /// Panics if the source status history collection is not registered in `introspection_ids`
+    pub async fn append_source_updates(&mut self, updates: Vec<RawStatusUpdate>) {
+        self.append_updates(updates, IntrospectionType::SourceStatusHistory)
+            .await;
+    }
+
+    /// Appends updates for sinks to the appropriate managed status collection
+    ///
+    /// # Panics
+    ///
+    /// Panics if the sink status history collection is not registered in `introspection_ids`
+    pub async fn append_sink_updates(&mut self, updates: Vec<RawStatusUpdate>) {
+        self.append_updates(updates, IntrospectionType::SinkStatusHistory)
+            .await;
+    }
+
+    async fn drop_items(&mut self, ids: Vec<GlobalId>, ts: EpochMillis, type_: IntrospectionType) {
+        self.append_updates(
+            ids.iter()
+                .map(|id| RawStatusUpdate {
+                    id: *id,
+                    ts,
+                    status_name: "dropped".to_string(),
+                    error: None,
+                    hints: Default::default(),
+                    namespaced_errors: Default::default(),
+                })
+                .collect(),
+            type_,
+        )
+        .await;
+        ids.into_iter().for_each(|id| {
+            self.previous_statuses.remove(&id);
```

Comment:
I think this should stay there because `previous_statuses` describes the set of statuses the user sees in `mz_{source|sink}_statuses`. If we ever wanted to do "streaming" cleanup of the status history (instead of on startup) then we would want this map to be accurate. Also less exceptions to explain

Reply:
this seems fine to me; we do have the problem where we never clean out dropped sources ever; ill add a todo